### ISSUE TITLE
False Positive Seafile Upload File

### ISSUE THREAD

**Issue by deniskonovalov64** [coreruleset/coreruleset#2396]
### Description
When loading a file of any size, ModSecurity gives an error "Inbound Anomaly Score Exceeded (Total Score: 5)". The solution provided in #2108 to increase the SecRequestBodyNoFilesLimit parameter did not give any result.

The only solution that I had to use is to disable the response to the error id by passing the `modsecurity_rules 'SecRuleRemoveById 949110'` command in the nginx config.

### Audit Logs / Triggered Rule Numbers
This is the data from the seafile error log:

``
2022/02/15 11:27:35 [error] 92281#92281: *10 [client -] ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `5' ) [file "/etc/nginx/modsec/coreruleset-nightly/rules/REQUEST-949-BLOCKING-EVALUATION.conf"] [line "139"] [id "949110"] [rev ""] [msg "Inbound Anomaly Score Exceeded (Total Score: 5)"] [data ""] [severity "2"] [ver "OWASP_CRS/3.4.0-dev"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [hostname "-"] [uri "/seafhttp/upload-aj/d27d43f6-dd71-4229-8b7c-e22c9d21936f"] [unique_id "1644884854"] [ref ""], client: -, server: -, request: "POST /seafhttp/upload-aj/d27d43f6-dd71-4229-8b7c-e22c9d21936f?ret-json=1 HTTP/1.1", host: "-", referrer: "-"
``

This is the data from the Mod Security audit log provided as JSON:
```
{
   "transaction":{
      "client_ip":"-",
      "time_stamp":"Tue Feb 15 11:27:34 2022",
      "server_id":"6190d41d7ace5d863bfe1641802c119a1867c446",
      "client_port":29125,
      "host_ip":"-",
      "host_port":443,
      "unique_id":"1644884854",
      "request":{
         "method":"POST",
         "http_version":1.1,
         "uri":"/seafhttp/upload-aj/d27d43f6-dd71-4229-8b7c-e22c9d21936f?ret-json=1",
         "headers":{
            "Content-Disposition":"attachment; filename=\"%D0%94%D0%BE%D0%BA%D1%83%D0%BC%D0%B5%D0%BD%D1%82%D1%8B.rar\"",
            "Content-Range":"bytes 0-8388607/13345862",
            "Content-Type":"multipart/form-data; boundary=----WebKitFormBoundaryy0EoxGj7AqAEKbM8",
            "User-Agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.82 Safari/537.36",
            "Sec-Fetch-Site":"same-origin",
            "sec-ch-ua-mobile":"?0",
            "Origin":"-",
            "Accept":"application/json; text/javascript, */*; q=0.01",
            "sec-ch-ua":"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"98\", \"Google Chrome\";v=\"98\"",
            "Content-Length":"8389992",
            "Connection":"keep-alive",
            "Sec-Fetch-Mode":"cors",
            "Host":"cloud.poseidon.su",
            "sec-ch-ua-platform":"\"Windows\"",
            "Referer":"-",
            "Sec-Fetch-Dest":"empty",
            "Accept-Encoding":"gzip, deflate, br",
            "Cookie":"-",
            "Accept-Language":"ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7"
         }
      },
      "response":{
         "http_code":403,
         "headers":{
            "Server":"nginx/1.21.4",
            "Date":"Tue, 15 Feb 2022 00:27:35 GMT",
            "Content-Length":"555",
            "Content-Type":"text/html",
            "Connection":"keep-alive"
         }
      },
      "producer":{
         "modsecurity":"ModSecurity v3.0.6 (Linux)",
         "connector":"ModSecurity-nginx v1.0.2",
         "secrules_engine":"Enabled",
         "components":[
            "OWASP_CRS/3.4.0-dev\""
         ]
      },
      "messages":[
         {
            "message":"HTTP header is restricted by policy (/content-range/)",
            "details":{
               "match":"Matched \"Operator `Within' with parameter `/proxy/ /lock-token/ /content-range/ /if/ /user-agentt/' against variable `TX:header_name_content-range' (Value: `/content-range/' )",
               "reference":"o0,19v555,19t:lowercaseo0,13v460,13t:lowercaseo0,12v377,12t:lowercaseo0,10v250,10t:lowercaseo0,14v722,14t:lowercaseo0,16v229,16t:lowercaseo0,6v688,6t:lowercaseo0,6v500,6t:lowercaseo0,9v153,9t:lowercaseo0,14v129,14t:lowercaseo0,10v106,10t:lowercaseo0,14v750,14t:lowercaseo0,4v82,4t:lowercaseo0,18v658,18t:lowercaseo0,7v793,7t:lowercaseo0,14v771,14t:lowercaseo0,15v887,15t:lowercaseo0,6v975,6t:lowercaseo0,15v922,15t:lowercaseo21,15",
               "ruleId":"920450",
               "file":"/etc/nginx/modsec/coreruleset-nightly/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf",
               "lineNumber":"1141",
               "data":"Restricted header detected: /content-range/",
               "severity":"2",
               "ver":"OWASP_CRS/3.4.0-dev",
               "rev":"",
               "tags":[
                  "application-multi",
                  "language-multi",
                  "platform-multi",
                  "attack-protocol",
                  "paranoia-level/1",
                  "OWASP_CRS",
                  "capec/1000/210/272",
                  "PCI/12.1"
               ],
               "maturity":"0",
               "accuracy":"0"
            }
         },
         {
            "message":"Inbound Anomaly Score Exceeded (Total Score: 5)",
            "details":{
               "match":"Matched \"Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `5' )",
               "reference":"",
               "ruleId":"949110",
               "file":"/etc/nginx/modsec/coreruleset-nightly/rules/REQUEST-949-BLOCKING-EVALUATION.conf",
               "lineNumber":"139",
               "data":"",
               "severity":"2",
               "ver":"OWASP_CRS/3.4.0-dev",
               "rev":"",
               "tags":[
                  "application-multi",
                  "language-multi",
                  "platform-multi",
                  "attack-generic"
               ],
               "maturity":"0",
               "accuracy":"0"
            }
         }
      ]
   }
}
```

### Your Environment

* CRS version (e.g., v3.2.0): 3.4.0-dev
* Paranoia level setting: default
* ModSecurity version (e.g., 2.9.3): v3.0.6
* Web Server and version (e.g., apache 2.4.41): nginx/1.21.4
* Operating System and version: Ubuntu 18.04.6 LTS

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by RedXanadu** [comment]
Hi @deniskonovalov64,

Thanks for the report!

First up, using `SecRuleRemoveById 949110` prevents the Core Rule Set from taking any blocking actions. _You almost certainly don't want to do that._ If you want CRS and ModSecurity to _scan only_, and _not_ to block any traffic, then either use `DetectionOnly` mode **or** set a very high inbound anomaly score threshold instead. No good can come from meddling with rule 949110, only confusion and potential bypass.

Now let's examine your log sample and see what the problem is.

`"message":"HTTP header is restricted by policy (/content-range/)",`
`"ruleId":"920450",`
`"data":"Restricted header detected: /content-range/",`

Rule 920450 blocks the use of certain "restricted" HTTP headers. CRS lists [Content-Range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range) as a restricted header, by default. If your application requires the use of the `Content-Range` header then you may want to remove it from the restricted headers list.

To do that, you need to edit your `crs-setup.conf` file. Find the section "Forbidden request headers". Uncomment rule 900250, like so:

```
SecAction \
 "id:900250,\
  phase:1,\
  nolog,\
  pass,\
  t:none,\
  setvar:'tx.restricted_headers=/proxy/ /lock-token/ /content-range/ /if/ /user-agentt/'"
```

Remove `/content-range/` from the list of restricted headers, like so:

```
  setvar:'tx.restricted_headers=/proxy/ /lock-token/ /if/ /user-agentt/'"
```

Then restart your web server to pick up the changes and retest your web application. It should then work!

If you have further issues then please reply here.

**Comment by deniskonovalov64** [comment]
Thanks to @RedXanadu, after removing the Content-Range from the forbidden headers, uploading files was possible, but deleting the file was followed by the same error.

This is the error log when uninstalling:

2022/02/17 09:13:08 [error] 11652#11652: *296 [client -] ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `5' ) [file "/etc/nginx/modsec/coreruleset-nightly/rules/REQUEST-949-BLOCKING-EVALUATION.conf"] [line "139"] [id "949110"] [rev ""] [msg "Inbound Anomaly Score Exceeded (Total Score: 5)"] [data ""] [severity "2"] [ver "OWASP_CRS/3.4.0-dev"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [hostname "-"] [uri "/api/v2.1/repos/batch-delete-item/"] [unique_id "1645049588"] [ref ""], client: -, server: -, request: "DELETE /api/v2.1/repos/batch-delete-item/ HTTP/1.1", host: "-", referrer: "-"

This is the ModSucurity audit log:
```
{
    "transaction": {
        "client_ip": "192.168.42.65",
        "time_stamp": "Thu Feb 17 09:13:08 2022",
        "server_id": "6190d41d7ace5d863bfe1641802c119a1867c446",
        "client_port": 24562,
        "host_ip": "-",
        "host_port": 443,
        "unique_id": "1645049588",
        "request": {
            "method": "DELETE",
            "http_version": 1.1,
            "uri": "/api/v2.1/repos/batch-delete-item/",
            "headers": {
                "Content-Type": "application/json;charset=UTF-8",
                "sec-ch-ua-mobile": "?0",
                "Origin": "-",
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36",
                "Sec-Fetch-Site": "same-origin",
                "X-CSRFToken": "qcR7MWN7Ce6mEcBcIVy5IFsyZlo1wtokD5FEVtwxSkjMn95Nz8J96JBjEYz4x7YU",
                "Accept-Encoding": "gzip, deflate, br",
                "Cookie": "sfcsrftoken=qcR7MWN7Ce6mEcBcIVy5IFsyZlo1wtokD5FEVtwxSkjMn95Nz8J96JBjEYz4x7YU; sessionid=fl3f0p9t5q3ofbghk8dr5jyx8v4c6hzy",
                "Accept": "application/json, text/plain, */*",
                "sec-ch-ua": "\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"98\", \"Google Chrome\";v=\"98\"",
                "sec-ch-ua-platform": "\"Windows\"",
                "Referer": "-",
                "Content-Length": "104",
                "Connection": "keep-alive",
                "Sec-Fetch-Mode": "cors",
                "Host": "-",
                "Sec-Fetch-Dest": "empty",
                "Accept-Language": "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7"
            }
        },
        "response": {
            "http_code": 403,
            "headers": {
                "Server": "nginx/1.21.4",
                "Date": "Wed, 16 Feb 2022 22:13:08 GMT",
                "Content-Length": "555",
                "Content-Type": "text/html",
                "Connection": "keep-alive"
            }
        },
        "producer": {
            "modsecurity": "ModSecurity v3.0.6 (Linux)",
            "connector": "ModSecurity-nginx v1.0.2",
            "secrules_engine": "Enabled",
            "components": ["OWASP_CRS/3.4.0-dev\""]
        },
        "messages": [{
                "message": "Method is not allowed by policy",
                "details": {
                    "match": "Matched \"Operator `Within' with parameter `GET HEAD POST OPTIONS' against variable `REQUEST_METHOD' (Value: `DELETE' )",
                    "reference": "v0,6",
                    "ruleId": "911100",
                    "file": "/etc/nginx/modsec/coreruleset-nightly/rules/REQUEST-911-METHOD-ENFORCEMENT.conf",
                    "lineNumber": "28",
                    "data": "DELETE",
                    "severity": "2",
                    "ver": "OWASP_CRS/3.4.0-dev",
                    "rev": "",
                    "tags": ["application-multi", "language-multi", "platform-multi", "attack-generic", "paranoia-level/1", "OWASP_CRS", "capec/1000/210/272/220/274", "PCI/12.1"],
                    "maturity": "0",
                    "accuracy": "0"
                }
            }, {
                "message": "Inbound Anomaly Score Exceeded (Total Score: 5)",
                "details": {
                    "match": "Matched \"Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `5' )",
                    "reference": "",
                    "ruleId": "949110",
                    "file": "/etc/nginx/modsec/coreruleset-nightly/rules/REQUEST-949-BLOCKING-EVALUATION.conf",
                    "lineNumber": "139",
                    "data": "",
                    "severity": "2",
                    "ver": "OWASP_CRS/3.4.0-dev",
                    "rev": "",
                    "tags": ["application-multi", "language-multi", "platform-multi", "attack-generic"],
                    "maturity": "0",
                    "accuracy": "0"
                }
            }
        ]
    }
}

```
As I understand it this time, he swears at the DELETE method, since it is not explicitly allowed, and in order for me to decide this, I should find the rule 911100 in crs-setup.conf, am I right?

**Comment by deniskonovalov64** [comment]
I found a mention of a similar problem in #1677 and added the DELETE method to the tx.allowed_methods variable in rule 900200, did I do everything right? Or can it be done in some other way?

**Comment by deniskonovalov64** [comment]
And another question, is it possible to write a log in JSON format, I don’t mean audit-log or access log, namely those logs that are written to the error log?

**Comment by deniskonovalov64** [comment]
Thanks, it helped me.  I close the question.
